### DISCUSSION TITLE
Update setting and addon privacy policy

### DISCUSSION THREAD

**Discussion by Kezxo** [tosdr/browser-extensions#58] [General]
I saw that this extension got an "update check" recently, I went to check the privacy policy on Mozilla but this didn't reflect that.

https://addons.mozilla.org/en-US/firefox/addon/terms-of-service-didnt-read/privacy/

It says 

> When the extension loads, it will retrieve one big file with all data from https://tosdr.org.

and

> Other than that, installing this extensions does not expose information to us or to any of the websites you visit.

Seems like this isn't up to date so it should be updated. Some things I would like the policy to address:

- When does it update?
- Are these update per site or is it that big file?
- If it's per site do you now track when people click the button? 
- I noticed that when i click the tosdr button with that update check disabled it seemingly still checks for updates. What is happening there? Is it downloading that huge icon in the popup?
- If I disable that update checkbox does that mean it will never ever contact your server again?

I think the privacy policy should be uploaded to github also so it can be read and tracked more easily.

If the addon downloads those icons from a server when we click the button that should be a thing you can disable, the UI should also reflect that the addon doesn't know if it's up to date or not.

**Comment by JustinBack** [comment]
Hi!

Those are good points you raise.

Currently the way the extension works is to pull said "big file" on start of the browser (https://api.tosdr.org/v1/all.json) this is to show the actual ratings and show notifications in your browser.

When opening the popup a request to the service's json file gets made `facebook.com` -> `https://api.tosdr.org/v1/service/facebook.json`

Additionally in an async operation, a request to the endpoint `https://api.tosdr.org/updatecheck/{version_string}` gets made which pulls the latest GitHub Tag and compares it using php's version_compare. (This can be disabled in the settings)

We also display shields data for the service which gets retrieved from `https://shields.tosdr.org/{service_id}.svg`
This can also be disabled in the settings)

4 Requests in total (-1 which gets made on browser startup)


As for the data collected.

We Cache our API files with a CDN which does not collect any information about you, not even the IP Address
(Source: I am the developer of said CDN)

![](https://9888-983528b48d7266437-14a87dc0-4ce4-11eb-b59c-5404a6a07802.imp.jbcdn.net/captures/mstsc_pbsvYOr60L.png)

Apache does not log any IP addresses anymore on the CDN. This was the case ~ a month ago however its really unnecessary and we periodically clean the log files completely so no IPs are collected anymore.

[(Exception is tosdr.org which uses your IP as an identifier to implement ratelimiting)](https://github.com/tosdr/CrispCMS/blob/6c0086e31fe83c0c20e731f78375be781d8f55c8/pixelcatproductions/crisp.php#L190-L200)

We use redis as a ratelimiting server.

![](https://9888-983528b48d7266437-14a87dc0-4ce4-11eb-b59c-5404a6a07802.imp.jbcdn.net/captures/mstsc_uIUIXYUJ1E.png)
(AccessFromIP is unused in the CDN)
![](https://9888-983528b48d7266437-14a87dc0-4ce4-11eb-b59c-5404a6a07802.imp.jbcdn.net/captures/mstsc_GCQjPG5K8H.png)

The only thing we count is the hits against our API, this excludes the cache

I'll make an example:

Everytime a request is made against `tosdr.org/api`, we increase the hit to the specific interface by one

![](https://9888-983528b48d7266437-14a87dc0-4ce4-11eb-b59c-5404a6a07802.imp.jbcdn.net/captures/mstsc_RJWWlhM4of.png)
We can see that an API request for facebook has been made, but not when and where

![](https://9888-983528b48d7266437-14a87dc0-4ce4-11eb-b59c-5404a6a07802.imp.jbcdn.net/captures/mstsc_luYo6Dz3zB.png)
(The login request is for our admin login api)


**However** since the extension goes to api.tosdr.org and the CDN caches that response, only 1 request every 1 hour is made to the actual endpoint. So we only count CDN hits basically.

![](https://9888-983528b48d7266437-14a87dc0-4ce4-11eb-b59c-5404a6a07802.imp.jbcdn.net/captures/mstsc_2qTn5Ix5E9.png)

> If I disable that update checkbox does that mean it will never ever contact your server again?

The updatecheck api does not get contacted, thats correct.
https://github.com/tosdr/browser-extensions/blob/dc7f086362dba390506dfdc93db85e895a025659/src/popup/js/popup.js#L123-L140

**BUT** the extension will still load service data from the server, this is required for the extension to work. Theoratically, the extension works offline, but really limited as in you cannot open the popup. The initial data on startup (the big file) gets cached so it will remain even without internet. This allows you to view ratings and get notifications in the url bar but opening the popup will fail.

(Same principle for shields, disabling this will not send a request to the server)
https://github.com/tosdr/browser-extensions/blob/dc7f086362dba390506dfdc93db85e895a025659/src/popup/js/popup.js#L90-L94

The Disable Update Check is only useful to stop checking if an extension update is available on your browser's store. **THIS DOES NOT DISABLE SERVICE WEB REQUESTS**

> If it's per site do you now track when people click the button?

Long answer above, short answer: Not really since requests are cached, only initial requests to the API. Not when and not who made it.

> When does it update?

Everytime you click on the popup

## If there are still questions feel free to ask!

**Reply by JustinBack** [reply]
I have now added the Privacy Policy in the PRIVACY.md file.

We will also deprecate the API Stats thing, its useless anyway as the CDN handles most of the work (this was used to balance requests across servers to handle the load in the past)

**Reply by JustinBack** [reply]
This has now been disabled in https://github.com/tosdr/CrispCMS/commit/6fd8aeff9a0a72c0b25597908b468d1bb43828eb

**Reply by Kezxo** [reply]
Thanks for the detailed response!

Hmm, I thought "the big file" contained all the info from the website and not just the ratings so you could click the popup even if tosdr went down for example. It's a bit more server based than I thought.

Now I know how it works at least, thanks again for the info and thanks for clarifying the privacy policy.

**Reply by JustinBack** [reply]
That was before we migrated to the new system, so your initial thought was correct 😄. However since we would have to periodically fetch data for each service (points, documents, ratings) the performance was REALLY REALLY bad and took up to 1 minute to complete. With the current system we only request files which are demanded and saves us a lot of resource.

One cool thing is, the extension has a small backwards compability with the old API. So you could clone yourself the now archived https://github.com/tosdr/tosdr.org/ repository (which has outdated data) and the extension would still work.

Technically you can host your own API and switch the endpoint using the override setting so that zero data is transmitted to us.
![](https://9888-983528b48d7266437-14a87dc0-4ce4-11eb-b59c-5404a6a07802.imp.jbcdn.net/captures/firefox_eJ1ZUQpdXy.png)

You could setup your own backend for instance and mirror our API.


> Hmm, I thought "the big file" contained all the info from the website

The "big file" contains lite data of *all* websites (urls, name, id, rating) which are required for the rating to show in your url bar, nothing more. In the old system, it did in fact, contain all data.

![](https://9888-983528b48d7266437-14a87dc0-4ce4-11eb-b59c-5404a6a07802.imp.jbcdn.net/captures/firefox_OKuO3YTHaD.png)


If there are any doubts you can setup your own Tos;DR environment by cloning our 3 main repositories and setting up a system yourself to test everything.

https://github.com/tosdr/browser-extensions
https://github.com/tosdr/edit.tosdr.org
https://github.com/tosdr/CrispCMS -> Docker: https://hub.docker.com/repository/docker/jbackpcat/tosdr-crisp/

Those 3 repos are our main system how everything works.
